browser/ui/tabs/shared_pinned_tab_service.cc

```diff
@@ -0,0 +1,680 @@
+    return;
+  }
+
+  MoveSharedWebContentsToBrowser(last_active_browser_, index);
+}
+
+void SharedPinnedTabService::MoveSharedWebContentsToBrowser(Browser* browser,
+                                                            int index) {
+  auto* tab_strip_model = browser->tab_strip_model();
+  DCHECK_LT(index, tab_strip_model->count());
+
+  LOCK_REENTRANCE(tab_strip_model);
+
+  auto& pinned_tab_data = pinned_tab_data_.at(index);
+  if (pinned_tab_data.contents_owner_model) {
+    std::unique_ptr<content::WebContents> unique_shared_contents;
+    unique_shared_contents =
+        pinned_tab_data.contents_owner_model->ReplaceWebContentsAt(
+            index, CreateDummyWebContents(pinned_tab_data.shared_contents));
+    DCHECK_EQ(pinned_tab_data.shared_contents, unique_shared_contents.get());
+    pinned_tab_data.contents_owner_model = tab_strip_model;
+
+    auto* dummy_contents =
+        pinned_tab_data.contents_owner_model->GetWebContentsAt(index);
+    auto* dummy_contents_data =
+        DummyContentsData::FromWebContents(dummy_contents);
+    DCHECK(dummy_contents_data);
+    dummy_contents_data->stop_propagation();
+
+    pinned_tab_data.contents_owner_model->ReplaceWebContentsAt(
+        index, std::move(unique_shared_contents));
+  } else {
+    // Restore a shared pinned tab from a closed browser.
+    auto iter =
+        base::ranges::find(cached_shared_contentses_from_closing_browser_,
+                           pinned_tab_data.shared_contents,
+                           &std::unique_ptr<content::WebContents>::get);
+    DCHECK(iter != cached_shared_contentses_from_closing_browser_.end());
+    auto unique_shared_contents = std::move(*iter);
+    DCHECK(unique_shared_contents);
+
+    cached_shared_contentses_from_closing_browser_.erase(iter);
+
+    pinned_tab_data.contents_owner_model = tab_strip_model;
+
+    auto* dummy_contents_data = DummyContentsData::FromWebContents(
+        tab_strip_model->GetWebContentsAt(index));
+    DCHECK(dummy_contents_data);
+    dummy_contents_data->stop_propagation();
+
+    // Unfortunately, We can't replace existing tab contents with cached web
+    // contents. We should use restore method.
+    chrome::AddRestoredTabFromCache(
+        std::move(unique_shared_contents), browser, index,
+        /* group= */ {},
+        /* select */ true, /* pin= */ true, /* user_agent_override= */ {},
+        /* extra_data= */ {});
+
+    // In order to prevent browser from being closed, we should close the dummy
+    // contents after we restore the tab.
+    tab_strip_model->CloseWebContentsAt(index + 1, /* close_type */ 0);
+  }
+}
+
+std::unique_ptr<content::WebContents>
+SharedPinnedTabService::CreateDummyWebContents(
+    content::WebContents* shared_contents) {
+  content::WebContents::CreateParams create_params(profile_);
+  create_params.initially_hidden = true;
+  create_params.desired_renderer_state =
+      content::WebContents::CreateParams::kNoRendererProcess;
```

Comment:
Check if we can share one site instance for all dummy web contents. cc @petemill

Reply:
Maybe, kNoRendererProcess make the process not forked at all?

Reply:
Yes for now there is no extra process